src/goto-programs/uncaught_exceptions.cpp

```diff
@@ -0,0 +1,306 @@
+
+class exceptiont
+{
+public:
+  typedef std::vector<std::pair<irep_idt,
+    goto_programt::targett>> catch_handlerst;
+  catch_handlerst catch_handlers;
+};
+
+typedef std::vector<exceptiont> stack_catcht;
+
+class uncaught_exceptionst
+{
+public:
+  explicit uncaught_exceptionst(
+    symbol_tablet &_symbol_table,
+    goto_functionst &_goto_functions):
+    symbol_table(_symbol_table),
+    goto_functions(_goto_functions)
+  {
+  }
+
+  void operator()(std::map<irep_idt, std::set<irep_idt>> &);
+
+protected:
+  symbol_tablet &symbol_table;
+  goto_functionst &goto_functions;
+
+  void collect_uncaught_exceptions(
+    std::map<irep_idt, std::set<irep_idt>> &);
+
+  void get_subtypes(
+    const irep_idt &,
+    std::vector<irep_idt> &);
+};
+
+/*******************************************************************\
+
+Function: uncaught_exceptionst::get_subtypes
+
+  Inputs:
+
+ Outputs:
+
+ Purpose: computes the set of subtypes of "type" 
+\*******************************************************************/
+
+void uncaught_exceptionst::get_subtypes(
+  const irep_idt &type,
+  std::vector<irep_idt> &subtypes)
+{
+  irep_idt candidate;
+  bool new_subtype=true;
+
+  // as we don't know the order types have been recorded,
+  // we need to iterate over the symbol table until there are no more
+  // new subtypes found
+  while(new_subtype)
+  {
+    new_subtype=false;
+    // iterate over the symbol_table in order to find subtypes of type
+    forall_symbols(it, symbol_table.symbols)
+    {
+      // we only look at type entries
+      if(it->second.is_type)
+      {
+        // every type is a potential candidate
+        candidate=it->second.name;
+        // current candidate is already in subtypes
+        if(find(subtypes.begin(),
+                subtypes.end(),
+                candidate)!=subtypes.end())
+          continue;
+        // get its base class
+        irept::subt bases=to_class_type((it->second).type).bases();
+        if(bases.size()>0)
+        {
+          irept base = bases[0];
+          const irept &base_type=base.find(ID_type);
+          assert(base_type.id()==ID_symbol);
+
+          // is it derived from type?
+          // i.e. does it have type or one of its subtypes as a base?
+          if(base_type.get(ID_identifier)==type ||
+             find(subtypes.begin(), subtypes.end(),
+                  base_type.get(ID_identifier))!=subtypes.end())
+          {
+            subtypes.push_back(candidate);
+            new_subtype=true;
+          }
+        }
+      }
+    }
+  }
+}
+
+/*******************************************************************\
+
+Function: uncaught_exceptionst::collect_uncaught_exceptions
+
+  Inputs:
+
+ Outputs:
+
+ Purpose: computes a map associating each function with 
+          an overapproximation of the set of exceptions that escape it
+
+\*******************************************************************/
+
+void uncaught_exceptionst::collect_uncaught_exceptions(
```

Comment:
I would have thought so as well; and in theory the escape analysis should already lay some foundations for achieving this?